feat: full cold compaction

carols10cents opened the pull request:
Closes #5330. Closes #5512.

This adds another stage to cold compaction. Currently, cold compaction will look for partitions with cold level 0 files and will compact them with any overlapping level 1 files into new level 1 files.

After this PR, cold compaction will look for partitions with cold level 0 *or level 1* files. It will still do the compaction of level 0 to level 1 as above, then it will take all the level 1 files in that partition, split them into groups if the size exceeds the max desired bytes, and then compact each group together into a level 2 file that won't be compacted again by either hot or cold compaction.

carols10cents commented:
Ok @pauldix I think this is ready for review now!

carols10cents commented:
@pauldix @NGA-TRAN I think this PR is ready for review, but it's big because I did some work on cold compaction before the memory budget, and then re-worked some of that work to re-use the memory budget code. So while I'd usually recommend reviewing a PR like this commit-by-commit, some of the later commits undo work of the earlier commits so you'd be reviewing code that doesn't end up in the final output.

However, I did a lot of moving around of code to share it between hot and cold compaction, so if you review the final diff, you'll be reviewing a lot of code that hasn't actually changed 😞 

Let me know if you'd like me to spend more time cleaning up the older commits to make this easier to review.

Also, I'm not sure that this is *done* done yet. Cold compaction used to work like this:

- Find cold compaction candidate partitions
- For a group of partitions determined to be able to run in parallel, for each partition:
  - Compact any remaining level 0 files to level 1
  - Group then compact all level 1 files to level 2 files, compacting each group in serial

Now it works like this:

- Find cold compaction candidate partitions
- For groups of candidate partitions and their level 0 + level 1 files chosen using the memory budget determined to be able to run in parallel, for each partition:
  - Compact any remaining level 0 files to level 1
- For each cold compaction candidate partition, in serial,
  - Group then compact all level 1 files to level 2 files, compacting each group in serial

I think it will be possible to reuse the memory budget code to run full compaction operations in parallel, but that would mean even more rearranging, and it felt like perhaps this was a good stopping point to make sure this is working and going in the right direction.

NGA-TRAN commented:
@carols10cents : I start reviewing the code but I think your serial point go well with my concern of OOM before reading this. 👍 for going simple first. Detail below.

> Now it works like this:

> Find cold compaction candidate partitions
> For groups of candidate partitions and their level 0 + level 1 files chosen using the memory budget determined to be able to run in parallel, for each partition:
Compact any remaining level 0 files to level 1
For each cold compaction candidate partition, in serial,
Group then compact all level 1 files to level 2 files, compacting each group in serial
I think it will be possible to reuse the memory budget code to run full compaction operations in parallel, but that would mean even more rearranging, and it felt like perhaps this was a good stopping point to make sure this is working and going in the right direction.

I totally agree with this. The `Group then compact all level 1 files to level 2 files` may need to use more memory if we have to  read more level 1 and level 2 files. Which means we  need a new memory budget  calculation for them. Let us start with serial compaction and when everything runs fine (we usually introduce new bugs with a new feature 😄 ), we will add parallelism for this

NGA-TRAN reviewed (commented):
What a PR 🎉 . This is great and did the work we want. There are 3 things I think we want to address before the merge:

1. When we `group_by_size` L1 files for compaction, we have to group only time-contiguous ones to keep the output result (L2s) non-overlapped.
2. In the `group-by_size`, we also want to compute memory needed to compact them to avoid OOM.
3. In `full_compaction`, we need to only compact L1s (to L2) that do not overlap with any L2s to keep all L2 non-overlapped. I think this can be done inside the function `group_by_size`, too.

NGA-TRAN commented on iox_catalog/src/postgres.rs:
```diff
@@ -1748,7 +1750,7 @@ limit $4;
 SELECT partition_id, shard_id, namespace_id, table_id, count(id), max(created_at)
 FROM   parquet_file
-WHERE  compaction_level = 0
+WHERE  (compaction_level = $4 OR compaction_level = $5)
```
Created ticket https://github.com/influxdata/conductor/issues/1152

carols10cents commented on clap_blocks/src/compactor.rs:
```diff
@@ -164,15 +120,18 @@ macro_rules! gen_compactor_config {
+            ///
+            /// The number of candidates compacted concurrently is also decided using this
+            /// estimation and budget.
+            ///
```
So should this not set a `default_value` in the clap attributes or should that value be different? Clap will still print out the `default_value` even if I remove the comment that explains what it is in GB.

NGA-TRAN replied:
I think we have to set the default and just remove comment. Yeah, clap still print out the default and we need this in the code. Let us not do anything and just keep the comment 😄

carols10cents commented on compactor/src/cold.rs:
```diff
@@ -0,0 +1,1031 @@
+        group_file_size_bytes += file.file_size_bytes() as u64;
+        group.push(file);
+
+        if group_file_size_bytes >= max_file_size_bytes {
```
By "beside this condition", do you mean instead of this condition or in addition to this condition?

NGA-TRAN replied:
"in addition to this condition"

NGA-TRAN commented on compactor/src/cold.rs:
```diff
@@ -0,0 +1,1031 @@
+    )
+    .await;
+
+    // Compact level 1 files to level 2 in serial (for now)
```
@carols10cents I have thought a lot about this over the weekend. I think it will be a lot easier for us isolate and narrow down the problem if we split this PR into 2. The first one only includes compacting L0s and L1s to L1s. The second one include `full_compaction` that compacts L1s into L2. We will have the first one merged and observer the production cluster for one or 2 days and if no major bugs found, we will merge the second one.
Each PR has its own and different complexity to keep data non-overlapped and we want to make sure they work as design.

carols10cents replied:
Ok sounds good, I can do that!

carols10cents commented on compactor/src/parquet_file_filtering.rs:
```diff
@@ -133,10 +134,11 @@ pub(crate) fn filter_hot_parquet_files(
     let ParquetFilesForCompaction {
         level_0,
         level_1: mut remaining_level_1,
+        .. // Ignore other levels
```
Correct!

carols10cents commented on compactor/src/compact.rs:
```diff
@@ -931,7 +929,7 @@ mod tests {
             row_count: 0,
             compaction_level: CompactionLevel::Initial, // level of file of new writes
-            created_at: time_38_hour_ago,               // create cold files by default
+            created_at: time_18_hour_ago,               // create cold files by default
```
I don't really see how it matters, but sure. I changed it to 9 hours so I didn't have to deal with floats

carols10cents commented on compactor/src/lib.rs:
```diff
@@ -47,69 +264,60 @@ pub(crate) enum Error {
-    let parquet_files_for_compaction =
-        parquet_file_lookup::ParquetFilesForCompaction::for_partition(
+    if to_compact.files.len() == 1
+        && to_compact.files[0].compaction_level() == CompactionLevel::Initial
```
> I wonder whether we ever hit len=1 and compaction_level=1 

If `filter_parquet_files` gets a group that contains 0 level 0 files and 1 level 1 file, it returns an empty `FilteredFiles` which won't get here

> (or even =2)

Level 2 shouldn't be possible here as `most_cold_files_partitions` only selects level 0 or level 1.

> But if we do, what should we do here? throw a warning?

I don't think we need to handle this specially? Calling `compact_parquet_files` on one level 1 file under the limits should just rewrite the file, which is doing a bit of extra work but shouldn't be making anything incorrect. And this situation should already be detectable looking for [this debug](https://github.com/influxdata/influxdb_iox/blob/762c2af91e07302a943a8c9eee84806325953af7/compactor/src/parquet_file_combining.rs#L130-L133), right?

carols10cents commented on compactor/src/parquet_file_combining.rs:
```diff
@@ -342,26 +513,7 @@ pub(crate) async fn compact_parquet_files(
-        compaction_input_file_bytes.record(size as u64);
-    }
-
-    Ok(())
```
I just started by copy pasting the whole method because I wasn't sure what all needed to be different. We can definitely refactor and unify them more in the future!

carols10cents commented:
Ok @NGA-TRAN I have disabled the cold full compaction and addressed comments that were relevant to the code that will be currently running-- comments about full compaction will be addressed in the next PR. Do you think this is ready to merge like this?

NGA-TRAN commented:
I am looking into this @carols10cents

NGA-TRAN reviewed (approved):
Without `full_compaction` invoked in this PR, I feel good to merge
